lib/frodata/concerns/api.rb

```diff
@@ -0,0 +1,262 @@
+      # attrs   - Hash of attributes to set on the record.
+      #
+      # Examples
+      #
+      #   # Update the lead with id '073ca9c8-2a41-e911-a81d-000d3a1d5a0b'
+      #   client.update('leads', "leadid": '073ca9c8-2a41-e911-a81d-000d3a1d5a0b', Name: 'Whizbang Corp')
+      #
+      # Returns true if the sobject was successfully updated.
+      # Returns false if there was an error.
+      def update(*args)
+        update!(*args)
+      rescue *exceptions
+        false
+      end
+
+      # Public: Update a record.
+      #
+      # entity_set - The set the entity belongs to
+      # attrs   - Hash of attributes to set on the record.
+      #
+      # Examples
+      #
+      #   # Update the leads with id '073ca9c8-2a41-e911-a81d-000d3a1d5a0b'
+      #   client.update!('leads', 'leadid' => '073ca9c8-2a41-e911-a81d-000d3a1d5a0b', "firstname" => 'Whizbang Corp')
+      #
+      # Returns true if the sobject was successfully updated.
+      # Raises an exception if an error is returned from Dynamics.
+      def update!(entity_set, attrs)
+        entity = service[entity_set].new_entity(attrs)
+        url_chunk = to_url_chunk(entity)
+
+        raise ArgumentError, 'ID field missing from provided attributes' if entity.is_new?
+
+        api_patch url_chunk, attrs
+        true
+      end
+
+      # Public: Delete a record.
+      #
+      # entity_set - The set the entity belongs to
+      # id      - The Dynamics primary key ID of the record.
+      #
+      # Examples
+      #
+      #   # Delete the lead with id  "073ca9c8-2a41-e911-a81d-000d3a1d5a0b"
+      #   client.destroy('leads',  "073ca9c8-2a41-e911-a81d-000d3a1d5a0b")
+      #
+      # Returns true if the entity was successfully deleted.
+      # Returns false if an error is returned from Dynamics.
+      def destroy(*args)
+        destroy!(*args)
+      rescue *exceptions
+        false
+      end
+
+      # Public: Delete a record.
+      #
+      # entity_set - The set the entity belongs to
+      # id      - The Dynamics primary key ID of the record.
+      #
+      # Examples
+      #
+      #   # Delete the lead with id  "073ca9c8-2a41-e911-a81d-000d3a1d5a0b"
+      #   client.destroy!('leads',  "073ca9c8-2a41-e911-a81d-000d3a1d5a0b")
+      #
+      # Returns true of the entity was successfully deleted.
+      # Raises an exception if an error is returned from Dynamics.
+      def destroy!(entity_set, id)
+        query = service[entity_set].query
+        url_chunk = query.find(id).to_s
+        api_delete url_chunk
+        true
+      end
+
+      # Public: Finds a single record and returns all fields.
+      #
+      # entity_set - The set the entity belongs to
+      # id      - The id of the record. If field is specified, id should be the id
+      #           of the external field.
+      # field   - External ID field to use (default: nil).
+      #
+      # Returns the Enitity record.
+      def find(entity_set, id, field = nil)
```

Comment:
Technically if you do have access to the EntitySet metadata you can get the field that is the key for the id.